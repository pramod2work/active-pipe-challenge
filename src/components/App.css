body {
  color: #425266;
  font-size: 12px;
  background-color: #6B7B8F;
}

.container {
  width: 100%;
  max-height: 100vh;
  display: -webkit-flex; /* Safari */
  display: flex;
  flex-wrap: wrap;
}

.cell {
  -webkit-flex-grow: 0; /* Safari 6.1+ */
  -webkit-flex-shrink: 0; /* Safari 6.1+ */
  -webkit-flex-basis: 46%; /* Safari 6.1+ */
  flex-grow: 0;
  flex-shrink: 0;
  flex-basis: 46%;
  margin: 1%;
  padding: 1%;
  background-color: white;
}

@media only screen and (min-width: 480px) {
  .cell{
    -webkit-flex-basis: 29.333%; /* Safari 6.1+ */
    flex-basis: 29.333%;
  }
}

@media only screen and (min-width: 992px) {
  .cell{
    -webkit-flex-basis: 16%; /* Safari 6.1+ */
    flex-basis: 16%;
  }
}