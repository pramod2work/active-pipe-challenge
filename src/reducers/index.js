import { combineReducers } from 'redux'

import listPropertiesData from './listProperties'

const rootReducer = combineReducers({
  listPropertiesData
})

export default rootReducer
