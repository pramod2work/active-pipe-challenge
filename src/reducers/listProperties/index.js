import { createAction, handleActions } from 'redux-actions'

// Actions
export const LIST_PROPERTIES_FETCH = 'LIST_PROPERTIES_FETCH'
export const LIST_PROPERTIES_SUCCESS = 'LIST_PROPERTIES_SUCCESS'
export const LIST_PROPERTIES_FAILURE = 'LIST_PROPERTIES_FAILURE'

const initialState = {
  data: null,
  error: false,
  inProgress: false,
}

export const listPropertiesFetch = createAction(LIST_PROPERTIES_FETCH)
export const listPropertiesSuccess = createAction(LIST_PROPERTIES_SUCCESS)
export const listPropertiesFailure = createAction(LIST_PROPERTIES_FAILURE)

const listPropertiesDataReducer = handleActions({
  [listPropertiesFetch]: (state, action) => ({
    ...initialState,
    inProgress: true
  }),
  [listPropertiesSuccess]: (state, action) => ({
    ...initialState,
    data: action.payload,
  }),
  [listPropertiesFailure]: (state, action) => ({
    ...initialState,
    error: action.payload,
  }),
}, initialState)

export default listPropertiesDataReducer
