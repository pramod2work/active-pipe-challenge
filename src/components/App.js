import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types'

import './App.css';
import { listPropertiesFetch } from '../reducers/listProperties';
import { PropertyInfo } from './PropertyInfo'
import { Filter } from './Filter'
import { Spinner } from './Spinner';
import { Error } from './Alert';

export class App extends Component {
  constructor(props) {
    super(props)

    this.state = {
      filter: ''
    }

    this.filterProperty = this.filterProperty.bind(this)
    this.getData = this.getData.bind(this)
  }

  componentDidMount() {
    this.getData()
  }

  filterProperty(filter) {
    this.setState({ filter })
  }

  getData() {
    this.props.listPropertiesFetch()
  }

  render() {
    return (
      <div className="App">
        {
          this.props.listPropertiesData.inProgress ?
            <Spinner />
          : ''
        }

        {
          this.props.listPropertiesData.error ?
            <Error retryAction={this.getData}/>
          : ''
        }

        {
          this.props.listPropertiesData.data ?
            <div>
              <Filter updateData={this.filterProperty} />
              <div className="container">
                {
                  this.props.listPropertiesData.data &&
                  this.props.listPropertiesData.data.map((property) => {
                    if (!this.state.filter || this.state.filter === property.status) {
                      return(
                        <PropertyInfo {...property} key={property.id} />
                      )
                    }

                    return null
                  })
                }
              </div>
            </div>
          : ''
        }
      </div>
    );
  }
}

App.defaultProps = {
  listPropertiesData: {
    data: false,
    inProgress: false,
    error: false
  },
  listPropertiesFetch: () => {}
}

PropertyInfo.propTypes = {
  listPropertiesData: PropTypes.shape({
    data: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.any
    ]),
    error: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.any
    ]),
    inProgress: PropTypes.bool,
  }),
  listPropertiesFetch: PropTypes.func,
}

export const mapStateToProps = ({
  listPropertiesData = {},
}) => ({
  listPropertiesData,
})

export const mapDispatchToProps = dispatch => (
  {
    listPropertiesFetch: () => {
      dispatch(listPropertiesFetch())
    },
  }
)

export default connect(mapStateToProps, mapDispatchToProps)(App);
