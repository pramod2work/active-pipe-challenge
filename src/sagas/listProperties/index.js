import { call, put } from 'redux-saga/effects'

import { getApiConfig, apiUriConfig } from '../../utils/apiConfig'
import { listPropertiesSuccess, listPropertiesFailure } from '../../reducers/listProperties'

export const getPropertiesAction = () => {
  const propertiesService = getApiConfig()
  return propertiesService.get(
    apiUriConfig.listProperties,
  )
}

export function* listProperties() {
  try {
    const response = yield call(getPropertiesAction)
    yield put(listPropertiesSuccess(response.data))
  } catch (error) {
    yield put(listPropertiesFailure(error.response))
  }
}
