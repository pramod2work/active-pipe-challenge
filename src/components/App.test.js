import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme'

import { App, mapStateToProps, mapDispatchToProps } from './App';
import * as listProperties from '../reducers/listProperties';

describe('App Component Testing', () => {
  test('renders without crashing', () => {
    const div = document.createElement('div');
    const props = {
      listPropertiesData: {
        data: [
          {
            "id": 1,
            "status": "current",
            "street": "1 Fake Street",
            "suburb": "Glen Iris",
            "state": "VIC",
            "postcode": "3146",
            "price": 1230500,
            "image": "https://code-challenge.activepipe.com/property-images/frontage/01.jpg"
          }
        ],
      },
      listPropertiesFetch: jest.fn()
    }
    ReactDOM.render(<App {...props} />, div);
    ReactDOM.unmountComponentAtNode(div);
    expect(props.listPropertiesFetch).toBeCalled()
  })

  test('Snapshot for Success scenario', () => {
    const props = {
      listPropertiesData: {
        data: [
          {
            "id": 1,
            "status": "current",
            "street": "1 Fake Street",
            "suburb": "Glen Iris",
            "state": "VIC",
            "postcode": "3146",
            "price": 1230500,
            "image": "https://code-challenge.activepipe.com/property-images/frontage/01.jpg"
          }
        ],
      },
      listPropertiesFetch: jest.fn()
    }
    const component = shallow(<App {...props} />)
    expect(component).toMatchSnapshot()
  })

  test('Snapshot for Error scenario', () => {
    const props = {
      listPropertiesData: {
        error: 'error',
      },
      listPropertiesFetch: jest.fn()
    }
    const component = shallow(<App {...props} />)
    expect(component).toMatchSnapshot()
  })

  test('Snapshot for Error scenario', () => {
    const props = {
      listPropertiesData: {
        inProgress: true,
      },
      listPropertiesFetch: jest.fn()
    }
    const component = shallow(<App {...props} />)
    expect(component).toMatchSnapshot()
  })

  test('on change of filter', () => {
    const props = {
      listPropertiesData: {
        data: [
          {
            "id": 1,
            "status": "current",
            "street": "1 Fake Street",
            "suburb": "Glen Iris",
            "state": "VIC",
            "postcode": "3146",
            "price": 1230500,
            "image": "https://code-challenge.activepipe.com/property-images/frontage/01.jpg"
          }
        ],
      },
      listPropertiesFetch: jest.fn()
    }
    const component = shallow(<App {...props} />)

    component.instance().filterProperty('invalid')
    expect(component.state().filter).toBe('invalid')
  })

  test('mapStateToProps for invalid state obj', () => {
    expect(mapStateToProps({})).toEqual({
      listPropertiesData: {},
    })
  })

  test('should return an object with the functions of dispatch event', () => {
    const dispatch = jest.fn()
    expect(mapDispatchToProps(dispatch)).toEqual({
      listPropertiesFetch: expect.any(Function),
    })
  })

  test('should call reducer action to fetch property data', () => {
    listProperties.listPropertiesFetch = jest.fn()
    const dispatch = jest.fn()
    const dispatchToProps = mapDispatchToProps(dispatch)

    dispatchToProps.listPropertiesFetch()
    expect(dispatch).toHaveBeenCalled()
    expect(listProperties.listPropertiesFetch).toBeCalled()
  })
})
