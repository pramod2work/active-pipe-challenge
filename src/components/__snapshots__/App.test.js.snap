// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component Testing Snapshot for Error scenario 1`] = `
<div>
  <Error
    retryAction={[Function]}
  />
</div>
`;

exports[`App Component Testing Snapshot for Error scenario 2`] = `
<div>
  <Spinner />
</div>
`;

exports[`App Component Testing Snapshot for Success scenario 1`] = `
<div>
  <div>
    <Filter
      updateData={[Function]}
    />
    <div
      className="container"
    >
      <PropertyInfo
        id={1}
        image="https://code-challenge.activepipe.com/property-images/frontage/01.jpg"
        key="1"
        postcode="3146"
        price={1230500}
        state="VIC"
        status="current"
        street="1 Fake Street"
        suburb="Glen Iris"
      />
    </div>
  </div>
</div>
`;
